File: Array/SubSet.cs
/*O(2^N)
Given a set of distinct integers, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,3], a solution is:
*/

public class Solution {
    public List<List<int>> Subsets(int[] nums) {
    
        List<int>list = new List<int>();
        List<List<int>>ans = new List<List<int>>();
        int size = nums.Length;
        Array.Sort(nums);
        for(int i = 0; i<(1<<size); i++){
            for(int j = 0; j<size; j++){
              if((i & (1<<j)) > 0){
                list.Add(nums[j]);
              }
            }
            ans.Add(list);
            list = new List<int>();
        }
        return ans;
    }
}

File: SubSet.cpp
/*
Given a set of distinct integers, S, return all possible subsets.

Note:

Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If S = [1,2,3], a solution is:

[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
*/
class Solution {
public:
    vector<vector<int> > subsets(vector<int> &S) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int sz = S.size(), i, j;
        vector<vector<int> >ans;
        vector<int>tmp; tmp.clear(); ans.clear();
        if(sz==0) return ans;
        sort(S.begin(),S.end());
        for(int i = 0; i<(1<<sz);i++)
         {
             for(int j = 0; j<sz; j++)
               if(i & (1<<j))
                 tmp.push_back(S[j]);
             ans.push_back(tmp);
             tmp.clear();
             
         }
        return ans;
    }
};